proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,267 @@
+#![deny(missing_docs)]
+use control::pb::telemetry::LatencyBucket;
+
+use std::{ops, slice, u32};
+use std::default::Default;
+use std::time::Duration;
+
+/// The number of buckets in a  latency histogram.
+pub const NUM_BUCKETS: usize = 26;
+
+/// A series of latency values and counts.
+#[derive(Debug)]
+pub struct Histogram([Bucket<Latency>; NUM_BUCKETS]);
+
+/// A latency in tenths of a millisecond.
+#[derive(Debug, Default, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Hash)]
+pub struct Latency(u32);
+
+/// A "leaf" bucket containing a single counter.
+///
+/// # Type Parameters:
+/// - `T`: the type of observations counted by this bucket.
+/// - `C`: the type used for counting observed values.
+#[derive(Debug, Clone)]
+pub struct Bucket<T, C = u32> {
+
+    /// The upper bound of the range of values in this bucket.
+    max: T,
+
+    /// Count of observations falling into this bucket.
+    count: C,
+
+}
+
+
+impl<T> Bucket<T> {
+
+    // Construct a new `LeafBucket` with the upper bound of `max`.
+    fn new<I: Into<T>>(max: I) -> Self {
+        Bucket {
+            max: max.into(),
+            count: 0
+        }
+    }
+
+}
+
+impl<T> Bucket<T>
+where
+    T: PartialOrd
+{
+
+    fn contains(&self, t: &T) -> bool {
+        t <= &self.max
+    }
+
+    fn observe(&mut self, t: &T) -> bool {
+        if self.contains(t) {
+            // silently truncate if the count is over u32::MAX.
+            self.count = self.count.saturating_add(1);
+            true
+        } else {
+            false
+        }
+
+    }
+}
+
+
+
+// ===== impl Histogram =====
+
+impl Histogram {
+
+    /// Observe a measurement
+    pub fn observe<I>(&mut self, measurement: I)
+    where
+        I: Into<Latency>,
+    {
+        let measurement = measurement.into();
+        for ref mut bucket in self {
+            if bucket.observe(&measurement) {
+                return;
+            }
+        }
+    }
+
+    /// Construct a new, empty `Histogram`.
+    ///
+    /// The buckets in this `Histogram` should mimic the Prometheus buckets
```

Comment:
I don't think this comment is accurate anymore?  We're not truncating, we're bucketing.

Reply:
We're still truncating an order of magnitude as well, though, right?

Reply:
I don't think there's any reason to truncate anymore.